cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java

```diff
@@ -0,0 +1,302 @@
+    }
+  }
+
+  private void updateAppProfileMetadata(ApplicationId applicationId, ApplicationSpecification appSpec) {
+    for (String name : appSpec.getWorkflows().keySet()) {
+      WorkflowId programId = applicationId.workflow(name);
+      updateProgramProfileMetadata(programId);
+    }
+  }
+
+  private void updateProgramProfileMetadata(ProgramId programId) {
+    ProfileId profileId = getResolvedProfileId(programId);
+    setProfileMetadata(programId, profileId);
+
+    for (ProgramSchedule schedule : scheduleDataset.listSchedules(programId)) {
+      updateScheduleProfileMetadata(schedule, profileId);
+    }
+  }
+
+  private void updateScheduleProfileMetadata(ProgramSchedule schedule, @Nullable ProfileId profileId) {
+    ScheduleId scheduleId = schedule.getScheduleId();
+    Optional<ProfileId> scheduleProfileId =
+      SystemArguments.getProfileIdFromArgs(scheduleId.getNamespaceId(), schedule.getProperties());
+    profileId = scheduleProfileId.orElse(profileId);
+    setProfileMetadata(scheduleId, profileId);
+  }
+
+  private void setProfileMetadata(NamespacedEntityId entityId, @Nullable ProfileId profileId) {
+    // if we are able to get profile from preferences or schedule properties, use it
+    // otherwise default profile will be used
+    metadataDataset.setProperty(entityId.toMetadataEntity(), PROFILE_METADATA_KEY,
+                                profileId == null ? ProfileId.DEFAULT.toString() : profileId.toString());
+  }
+
+  /**
+   * Get the profile id for the provided entity id from the resolved preferences from preference dataset
+   *
+   * @param entityId entity id to lookup the profile id
+   * @return the profile id which will be used by this entity id, null if not find
+   */
+  // TODO: CDAP-13579 consider preference key starts with [scope].[name].system.profile.name
+  @Nullable
+  private ProfileId getResolvedProfileId(EntityId entityId) {
+    if (entityId.getEntityType().equals(EntityType.INSTANCE)) {
+      return SystemArguments.getProfileIdFromArgs(NamespaceId.SYSTEM, getPreferences(entityId)).orElse(null);
+    }
+
+    if (entityId instanceof NamespacedEntityId) {
+      NamespaceId namespaceId = ((NamespacedEntityId) entityId).getNamespaceId();
+      Optional<ProfileId> profile =
+        SystemArguments.getProfileIdFromArgs(namespaceId, getPreferences(entityId));
+
+      // if we are able to find the profile, return it
+      if (profile.isPresent()) {
+        return profile.get();
+      }
+
+      // if the entity id has a parent id, get the preference from its parent
+      if (entityId instanceof ParentedId) {
+        return getResolvedProfileId(((ParentedId) entityId).getParent());
+      } else {
+        // otherwise it is a namespace id, which we want to look at the instance level
+        return getResolvedProfileId(new InstanceId(""));
+      }
+    }
+    return null;
+  }
+
+  private Map<String, String> getPreferences(EntityId entityId) {
```

Comment:
don't need this method, just call the dataset directly

Reply:
Removed